### PR TITLE
[android] Bump compileSdkVersion and targetSdkVersion to 34

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes ENG-10419
Bumps the compile and target SDK version to 34.

# How

Behavior changes:

- https://developer.android.com/about/versions/14/behavior-changes-all
- https://developer.android.com/about/versions/14/behavior-changes-14

Required Changes:

- `expo-location` needs to use the `Geocoder` directly as `SmartLocation` uses a context registered receiver and doesn't pass the required parameters causing android 14 devices to crash with a `SecurityException` note: we should remove [this](https://github.com/mrmans0n/smart-location-lib) dependency. It hasn't had a commit in 6 years
- `flashlist` needed a patch until they land [this](https://github.com/Shopify/flash-list/pull/930)
- bump installreferrer in `expo-application` because devices without access to google play services will hard crash.

# Test Plan

- Expo go with NCL and test-suite
  - emulator with API 34 ✅
  - emulator with API 31 ✅
  - physical device with API 34 ✅
- Bare expo
  - emulator with API 34 ✅
  - emulator with API 31 ✅
  - physical device with API 34 ✅
- Dev-Client
  - emulator with API 31 ✅
  - physical device with API 34 ✅

### CODE REVIEW

**Comment by reichhartd** [comment]
Just as a heads up, https://github.com/expo/expo/pull/24777 has been merged and `expo-tracking-transparency` now also supports Android. The package should be included in this PR.

File: packages/expo-splash-screen/android/build.gradle [hunk 1]
```diff
@@ -94,7 +94,7 @@ android {
   namespace "expo.modules.splashscreen"
   defaultConfig {
     minSdkVersion safeExtGet('minSdkVersion', 21)
-    targetSdkVersion safeExtGet("targetSdkVersion", 33)
+    targetSdkVersion safeExtGet("targetSdkVersion", 34)
```
Comment (Kudo) [inline packages/expo-splash-screen/android/build.gradle:97]:
```suggestion
```
nit: maybe we can remove this line

File: packages/expo-video/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,9 @@
 
 ### 💡 Others
 
+- Ship untranspiled JSX to support custom handling of `jsx` and `createElement`. ([#24889](https://github.com/expo/expo/pull/24889) by [@EvanBacon](https://github.com/EvanBacon))
```
Comment (Kudo) [inline packages/expo-video/CHANGELOG.md:13]:
```suggestion
```

unrelated?

File: packages/expo-application/CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,7 @@
 ### 💡 Others
 
 - Remove `unimodule.json` in favour of `expo-module.config.json`. ([#25100](https://github.com/expo/expo/pull/25100) by [@reichhartd](https://github.com/reichhartd))
+- On `Android` bump `compileSdkVersion` and `targetSdkVersion` to `34`. ([#24708](https://github.com/expo/expo/pull/24708) by [@alanjhughes](https://github.com/alanjhughes))
```
Comment (Kudo) [inline packages/expo-application/CHANGELOG.md:16]: would like to move those changelogs into "breaking changes" section. that would get more attention for the changes.

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
   "scripts": {
     "setup:docs": "./scripts/download-dependencies.sh",
     "setup:native": "./scripts/download-dependencies.sh && ./scripts/setup-react-android.sh",
-    "postinstall": "yarn-deduplicate && expo-yarn-workspaces check-workspace-dependencies && yarn workspace @expo/cli prepare",
+    "postinstall": "yarn-deduplicate && expo-yarn-workspaces check-workspace-dependencies && yarn workspace @expo/cli prepare && patch-package",
```
Comment (Kudo) [inline package.json:10]: nit * 3:
personally if i only have just one patch and that's temporary, i would prefer not to add patch-package and introduce extra dependency to the monorepo.
something like the reverse of #19516 was what i did before.

**Review by gabrieldonadel — APPROVED** [review]

File: yarn.lock [hunk 5]
```diff
@@ -6924,9 +6924,9 @@ ci-info@^2.0.0:
   integrity sha512-5tK7EtrZ0N+OLFMthtqOj4fI2Jeb88C4CAZPu25LDVUgXJ0A3Js4PMGqrn0JU1W0Mh1/Z8wZzYPxqUrXeBboCQ==
 
 ci-info@^3.2.0, ci-info@^3.3.0:
-  version "3.3.0"
-  resolved "https://registry.yarnpkg.com/ci-info/-/ci-info-3.3.0.tgz#b4ed1fb6818dea4803a55c623041f9165d2066b2"
-  integrity sha512-riT/3vI5YpVH6/qomlDnJow6TBee2PBKSEpx3O32EGPYbWGIRsIlGRms3Sm74wYE1JMo8RnO04Hb12+v1J5ICw==
+  version "3.9.0"
```
Comment (Kudo) [inline yarn.lock:6927]: not sure how yarn.lock changed since you didn't change any package.json. anyway the lockfile change looks good still.

**Review by Kudo — APPROVED** [review]
thanks for the help to fix sqlite/next module on android expo-go 💛
